src/PowerShellEditorServices/Services/PowerShell/Execution/BlockingConcurrentDeque.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.PowerShell.EditorServices.Services.PowerShell.Execution
+{
+    /// <summary>
+    /// Implements a concurrent deque that supplies:
+    ///   - Non-blocking prepend and append operations
+    ///   - Blocking and non-blocking take calls
+    ///   - The ability to block consumers, so that <see cref="Prepend(T)"/> can also guarantee the state of the consumer
+    /// </summary>
+    /// <typeparam name="T">The type of item held by this collection.</typeparam>
+    /// <remarks>
+    /// The prepend/append semantics of this class depend on the implementation semantics of <see cref="BlockingCollection{T}.TryTakeFromAny(BlockingCollection{T}[], out T)"/>
+    /// and its overloads checking the supplied array in order.
+    /// This behavior is unlikely to change and ensuring its correctness at our layer is likely to be costly.
+    /// See https://stackoverflow.com/q/26472251.
+    /// </remarks>
+    internal class BlockingConcurrentDeque<T>
```

Comment:
Hopefully this class is fairly self-explanatory. It's quite useful for our purposes. It doesn't work too hard to reimplement a new concurrent data structure, but instead takes advantage of the `BlockingCollection<T>.TryTakeFromAny()` method that .NET provides.

All the actual cleverness is still handled by [the `BlockingCollection<T>` implementation](https://github.com/dotnet/runtime/blob/27536d59df9b9e7263d48dfe910b48e3792c57a9/src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/BlockingCollection.cs#L659-L765), where consumers and producers just hold on to semaphores that track the capacity of the underlying collection.

Reply:
I don't see where this is used anywhere in this PR.
Also, should this not have a Dispose() method to clean up collections, event objects?
It is not clear to me what this is for, it would be nice to have comments illustrating its use.